lib/mutations/dependent_validations.rb

```diff
@@ -0,0 +1,108 @@
+    #     validate_country_code
+    #     validate_email
+    #   end
+    #
+    #   def validate_country_code
+    #     return unless user.country_code.blank?
+    #
+    #     add_error(:country, :not_provided, "You must provide your location first")
+    #   end
+    #
+    #   def validate_email
+    #     return if user.country_code.blank?
+    #     return if email.tld == user.country_code
+    #
+    #     add_error(:email, :mismatching_tld, "Your email's TLD must match your country code")
+    #   end
+    # end
+    #
+    # But this can quickly become unwieldy as validations further depend on each other. Failing to
+    # add the correct guards can result in confusing error messages (in this case, telling the user
+    # their email must match their country code, which they never provided).
+    #
+    # With this module, this becomes:
+    #
+    # class UpdateEmail < Mutations::Command
+    #   include Mutations::DependentValidations
+    #
+    #   def validations
+    #     {
+    #       validate_country_code:[],
+    #       validate_email: %i[validate_country_code],
+    #     }
+    #   end
+    #
+    #   def validate_country_code
+    #     return unless user.country_code.blank?
+    #
+    #     add_error(:country, :not_provided, "You must provide your location first")
+    #   end
+    #
+    #   def validate_email
+    #     return if email.tld == user.country_code
+    #
+    #     add_error(:email, :mismatching_tld, "Your email's TLD must match your country code")
+    #   end
+    # end
+    #
+    # Importantly, note that we no longer invert #validate_country_code's guards in #validate_email,
+    # nor do we provide an implementation of #validate, which is provided by the module.
+
+    CyclicDependencyError = Class.new(StandardError)
+    InvalidValidationsError = Class.new(StandardError)
+    UndefinedValidationsError = Class.new(StandardError)
+    UndeclaredValidationError = Class.new(StandardError)
+
+    def validate
+      raise UndefinedValidationsError, "No validations provided (as .validations class method)" unless self.class.respond_to?(:validations, true)
+
+      validations = self.class.validations
+      raise InvalidValidationsError, "All validations and dependencies must be symbols" unless validations.all? { |k,v| k.class == Symbol && v.all? { |dep| dep.class == Symbol } }
+
+      undeclared_validations = validations.values.flatten.uniq - validations.keys
+      raise UndeclaredValidationError, "Undeclared validations: #{undeclared_validations}, all dependencies must be explicit" if undeclared_validations.any?
+
+      validation_outcomes = {}
+      validations.keys.each do |validation|
+        # This executes validations and their dependencies, depth-first. In the example above, we
+        # would always run the address check first, regardless of the order in which .validations
+        # defines the dependencies.
+        recursively_run_validation(validation, validation_outcomes, Set.new)
```

Comment:
This does mean we rediscover the same ordering in which we run the validations every time, which is a bit inefficient. I originally wrote this to determine the ordering ahead of time and save it in an instance variable on the class, but it was much more complicated and felt like an unnecessary early optimisation.